server/core/src/replica_state_table.cpp

```diff
@@ -0,0 +1,237 @@
+
+        // Global Variables
+        nlohmann::json replica_state_map;
+
+        std::mutex rst_mutex;
+
+        auto index_of(std::string_view _logical_path, const int _replica_number) -> int
+        {
+            // NOTE: This function is not inherently thread-safe -- make sure to acquire the lock!
+
+            int index = -1;
+
+            for (const auto& e : replica_state_map.at(_logical_path.data()).at("replicas")) {
+                ++index;
+                if (_replica_number == std::stoi(e.at("before").at("data_repl_num").get<std::string>())) {
+                    return index;
+                }
+            }
+
+            THROW(KEY_NOT_FOUND, fmt::format(
+                "[{}:{}] - replica number [{}] not found for [{}]",
+                __FUNCTION__, __LINE__, _replica_number, _logical_path));
+        } // index_of
+    } // anonymouse namespace
+
+    auto replica_state_table::init() -> void
+    {
+        std::scoped_lock rst_lock{rst_mutex};
+
+        irods::log(LOG_DEBUG9, fmt::format("[{}:{}] - initializing state table", __FUNCTION__, __LINE__));
+
+        replica_state_map.clear();
+    } // init
+
+    auto replica_state_table::deinit() -> void
+    {
+        std::scoped_lock rst_lock{rst_mutex};
+
+        irods::log(LOG_DEBUG9, fmt::format("[{}:{}] - de-initializing state table", __FUNCTION__, __LINE__));
+
+        replica_state_map.clear();
+    } // deinit
+
+    auto replica_state_table::instance() noexcept -> replica_state_table&
+    {
+        static replica_state_table instance;
+
+        irods::log(LOG_DEBUG9, fmt::format("[{}:{}] - retrieving instance of replica state table", __FUNCTION__, __LINE__));
+
+        return instance;
+    } // instance
+
+    auto replica_state_table::insert(const DataObjInfo& _obj) -> void
+    {
+        std::string_view logical_path = _obj.objPath;
+
+        std::scoped_lock rst_lock{rst_mutex};
+
+        if (replica_state_map.contains(logical_path.data())) {
+            irods::log(LOG_DEBUG, fmt::format("[{}:{}] - entry exists;path:[{}]", __FUNCTION__, __LINE__, logical_path));
+            return;
+        }
+
+        const auto obj = irods::experimental::data_object::make_data_object_proxy(_obj);
+
+        nlohmann::json entry{{"data_id", std::to_string(obj.data_id())}};
+
+        for (const auto& r : obj.replicas()) {
+            const auto json_replica = irods::experimental::replica::to_json(r);
+
+            entry["replicas"].push_back({
+                {"before", json_replica},
+                {"after", json_replica}
+            });
+        }
+
+        replica_state_map[obj.logical_path().data()] = entry;
+    } // insert
+
+    auto replica_state_table::erase(std::string_view _logical_path) -> void
+    {
+        std::scoped_lock rst_lock{rst_mutex};
+
+        if (replica_state_map.contains(_logical_path.data())) {
+            replica_state_map.erase(_logical_path.data());
+        }
```

Comment:
Does this actually work? The lock is already acquired before calling `contains`. It should deadlock.

See https://gcc.godbolt.org/z/j7ErPY

Reply:
Yes, it works. This is `nlohmann::json::contains`, not `irods::replica_state_table::contains`.

Reply:
Excellent.